Add FireFly Graduation Proposal (#11)
ryjones/hyperledger-hip: HIPs/firefly_graduation.md

# Hyperledger FireFly Graduation Proposal

## Abstract

FireFly began as a [Hyperledger Lab in June, 2021](https://wiki.hyperledger.org/display/TSC/Table+consideration+of+the+FireFly+proposal+for+two+months) and became an [Incubating Project in September, 2021](https://wiki.hyperledger.org/display/TSC/Move+Hyperledger+Lab+Firefly+into+a+Project+in+Incubation). Now, after two years in Incubation FireFly has significantly grown and continued to mature. The project has received significant contributions from committers beyond the set of original maintainers, including a completely new [blockchain connector microservice](https://github.com/hyperledger/firefly-tezosconnect) and corresponding FireFly Core plugin enabling connectivity to Tezos, an entirely new type of blockchain. Additionally, the project now has a diverse set of maintainers representing three different companies.

> Given that FireFly now meets or exceeds the [Incubation Exit Criteria](https://toc.hyperledger.org/governing-documents/project-incubation-exit.html) created by the Hyperledger TOC, is a proposal to move FireFly to Graduated Project status. The rest of this document describes in detail how the project meets or exceeds the Incubation Exit Criteria.

Reviewer: The grammar in the first sentence seems off. Maybe
```suggestion
Given that FireFly now meets or exceeds the [Incubation Exit Criteria](https://toc.hyperledger.org/governing-documents/project-incubation-exit.html) created by the Hyperledger TOC, we are proposing to move FireFly to Graduated Project status. The rest of this document describes in detail how the project meets or exceeds the Incubation Exit Criteria.
```

Author: Thanks. Yep that looks like a sentence I edited poorly.

> Project name has been checked for trademark issues

Both of these requirements were met before FireFly moved in the Hyperledger Labs.

### Community support

> The project must have an active and diverse set of contributing members representing various constituencies

FireFly has a very active Discord channel where community members from all over the world regularly chat about project ideas, suggest improvements, ask technical questions, or seek troubleshooting assistance.

The community holds public Community Calls at least once a month where maintainers and contributors discuss architectural proposals, upcoming releases, and demo new features in the project.

> The project is not highly dependent on any single contributor (there are at least 3 legally independent committers and there is no single company or entity that is vital to the success of the project)

#### Committers

According to [LFX Insights](https://insights-v2.lfx.linuxfoundation.org/firefly/technical-contributors/participating-organizations?selectedDateFilterType=DATERANGE&selectedDateRangeKey=3Y), FireFly has committers representing 8 corporations and 12 independent contributors.

#### Maintainers

> FireFly has 12 maintainers representing three different companies:

Reviewer: With 10 maintainers coming from Kaleido, does this meet the `there is no single company or entity that is vital to the success of the project`? In other words, if Kaleido decided to stop supporting FireFly development, would the project continue to succeed? What is being done to move to a more diverse set of maintainers?

Reviewer 2: Specifically, the non-Kaleido maintainers appear to be active in the connectors, have they started to maintain the core Firefly repositories as well?

Author: > What is being done to move to a more diverse set of maintainers?

This is a good question. I'm happy to elaborate on this more in the TOC call tomorrow.

> Specifically, the non-Kaleido maintainers appear to be active in the connectors, have they started to maintain the core Firefly repositories as well?

Here are @denisandreenko's commits in FireFly Core: https://github.com/hyperledger/firefly/commits?author=denisandreenko (Many of these were made earlier in the year but the branch was rebased before it got merged.)

I will also make the case that contributions to connectors should not be considered "second class" to contributions to the main FireFly repo. We call the main repo "FireFly Core" because it is the piece that all other pieces connect to. But the "Core" without connectors is useless. Additionally, the connectors contain the some of the most difficult, blockchain-specific code in the project.

- [Kaleido](https://kaleido.io/)
- [Fidelity](https://www.fidelity.com/)
- [OneOf](https://www.oneof.com/)

It is worth noting that each company has at least one maintainer who is a subject matter expert on one or more specific blockchain connectors, and the most foundational parts of the FireFly technology stack.

For the full list of maintainers, please see the [FireFly Maintainers page](https://wiki.hyperledger.org/display/FIR/Maintainers) in the Hyperledger Wiki.

#### Steering Committee

The FireFly Community is in active discussions of establishing a Steering Committee for the project itself. This would be a role for stakeholders who have a vested interest in the direction and the success of the project but are not necessarily writing or reviewing code on a daily basis. We have several Hyperledger Member companies that have expressed interest in being involved in the project in such a way including:

- [DTCC](https://www.dtcc.com/)
- [OneOf](https://www.oneof.com/)
- [Kaleido](https://kaleido.io/)

#### Maintainer Pipeline

> The FireFly Community is actively working to continue to grow the FireFly technical contributors and maintainers through a Maintainer Pipeline. Pipeline activities range from working with organizations such as Grace Hopper Celebration to host open source contribution days, maintaining an active Discord channel, regular community calls, regular FireFly technical workshops, active plans to develop education/ workshop materials with local HL chapters and industry partners such as AWS, along with interactions with technical communities at major web3 community events such as EthDenver, EthCC, Consensus, Devconnect, etc. These pipeline activities aim to cultivate more technical contributions and maintainers from new organizations. We are pleased with the current continued growth in diversity in the project, and this "maintainer pipeline" will allow it to continue growing. Some examples of companies/organizations in various stages of this pipeline:

Reviewer: Can you comment on the `various stages` and where you see each of these organizations at in the pipeline?

Author: Yes! I'm planning on sharing more on this tomorrow during the call

- [DTCC](https://www.dtcc.com/)
- [LACChain](https://lacnet.lacchain.net/firefly/)
- [CGI Federal](https://www.cgi.com/us/en-us/federal)

> Release plans are developed and executed in public by the community.

Upcoming FireFly feature releases are tracked with [GitHub Milestones](https://github.com/hyperledger/firefly/milestones) with individual work items being tracked in [GitHub Issues](https://github.com/hyperledger/firefly/issues). Releases are also discussed during regular Community Calls.

### Sufficient test coverage

> The project must include a comprehensive unit and integration test suite and document its coverage. Additional performance and scale test capability is desirable.

> - FireFly Core has 100% unit test coverage which is enforced by Codecov [![codecov](https://codecov.io/gh/hyperledger/firefly/branch/main/graph/badge.svg?token=QdEnpMqB1G)](https://codecov.io/gh/hyperledger/firefly)

Reviewer: Do the other repositories also have unit tests? If so, what are their code coverage?

Author: Yes, they do 🙂 I will list the other repositories here as well

- FireFly has an End-to-End test suite that runs against a live system in a robust matrix of possible configurations
- Both the unit test and E2E test suites run against every PR

> - The [FireFly Performance CLI](https://github.com/hyperledger/firefly-perf-cli) is used for testing transaction throughput of a live system in various configurations

Reviewer: Is running the performance testing a one-off? Or is there data available for each of the releases on a given set of configurations? (Just curious...not a requirement for graduation.)

Author: Archiving performance runs for previous releases is something we would like a better system for. Until now, when we're working on testing a release candidate we have bene creating a thread for the release in Discord. In that thread, we dump all of the settings that we're using for the performance test as well as results. It works in the moment, but it's not very accessible afterward. We're open to suggestions if other projects have found good solutions for this sort of thing.

- AscendBit - AscendBit is a company owned by CP Group that is using Hyperledger FireFly to launch supply chain tracking initiatives, as well as its own company coin.
- LACChain is a blockchain infrastructure project designed for enabling inclusive and scalable projects and solutions on web 3.0 for Latin American and the Caribbean. FireFly is a key component of their tech stack and they have written up a tutorial for deploying FireFly onto LACChain.

It is worth noting that this is a list of projects that are publicly known to be using FireFly in production. There are additional significant proprietary projects that are not able to be listed here at this time.

### Sufficient scalability

> The project must demonstrate sufficient scalability and document its scalability over various dimensions

#### Transaction Throughput

Because FireFly is a modular, pluggable microservice architecture, there are an almost infinite number of variables involved in measuring or tuning performance and scalability. The biggest factor in transaction throughput usually ends up being the blockchain itself.

> Independent testing by separate companies has concluded that FireFly does not create a bottleneck on transaction throughput above that of the blockchain itself.

Reviewer: Curious if you could point to the results of these independent tests.

Author: I would love to be able to, but this testing was done in the context of an NDA.
